File: paging/lib/src/test/java/org/signal/paging/DataStatusTest.java
package org.signal.paging;

import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class DataStatusTest {

  @Test
  public void insertState_initiallyEmpty_InsertAtZero() {
    DataStatus subject = DataStatus.obtain(0);
    subject.insertState(0, true);

    assertEquals(1, subject.size());
    assertTrue(subject.get(0));
  }

  @Test
  public void insertState_someData_InsertAtZero() {
    DataStatus subject = DataStatus.obtain(2);
    subject.mark(1);

    subject.insertState(0, true);

    assertEquals(3, subject.size());
    assertTrue(subject.get(0));
    assertFalse(subject.get(1));
    assertTrue(subject.get(2));
  }

  @Test
  public void insertState_someData_InsertAtOne() {
    DataStatus subject = DataStatus.obtain(3);
    subject.mark(1);

    subject.insertState(1, true);

    assertEquals(4, subject.size());
    assertFalse(subject.get(0));
    assertTrue(subject.get(1));
    assertTrue(subject.get(2));
    assertFalse(subject.get(3));
  }

  @Test(expected = IndexOutOfBoundsException.class)
  public void insertState_negativeThrows() {
    DataStatus subject = DataStatus.obtain(0);
    subject.insertState(-1, true);
  }

  @Test(expected = IndexOutOfBoundsException.class)
  public void insertState_largerThanSizePlusOneThrows() {
    DataStatus subject = DataStatus.obtain(0);
    subject.insertState(2, true);
  }
}

File: lib/paging/src/test/java/org/signal/paging/DataStatusTest.kt
package org.signal.paging

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class DataStatusTest {
  @Test
  fun insertState_initiallyEmpty_InsertAtZero() {
    val subject = DataStatus.obtain(0)
    subject.insertState(0, true)

    assertEquals(1, subject.size())
    assertTrue(subject[0])
  }

  @Test
  fun insertState_someData_InsertAtZero() {
    val subject = DataStatus.obtain(2)
    subject.mark(1)

    subject.insertState(0, true)

    assertEquals(3, subject.size())
    assertTrue(subject[0])
    assertFalse(subject[1])
    assertTrue(subject[2])
  }

  @Test
  fun insertState_someData_InsertAtOne() {
    val subject = DataStatus.obtain(3)
    subject.mark(1)

    subject.insertState(1, true)

    assertEquals(4, subject.size())
    assertFalse(subject[0])
    assertTrue(subject[1])
    assertTrue(subject[2])
    assertFalse(subject[3])
  }

  @Test(expected = IndexOutOfBoundsException::class)
  fun insertState_negativeThrows() {
    val subject = DataStatus.obtain(0)
    subject.insertState(-1, true)
  }

  @Test(expected = IndexOutOfBoundsException::class)
  fun insertState_largerThanSizePlusOneThrows() {
    val subject = DataStatus.obtain(0)
    subject.insertState(2, true)
  }
}
